File: js/Grid.js
'use strict';

var main = require('./main');
var CONST = require('constants');

var GRID_BORDER = CONST.GRID_BORDER;


/** @class A generic grid.
 *  NB: We keep extra "border" cells around the real board.
 *      Idea is to avoid checking i,j against gsize in many places.
 *  public read-only attribute: gsize
 *  public RW attribute: yx
 */
function Grid(gsize, initValue, borderValue) {
    if (initValue === undefined) throw new Error('Grid init value must be defined');
    this.gsize = gsize;
    if (borderValue === undefined) {
        this.yx = main.newArray2(gsize + 2, gsize + 2, initValue);
    } else {
        this.yx = main.newArray2(gsize + 2, gsize + 2, borderValue);
        this.init(initValue);
    }
}
module.exports = Grid;

Grid.COLOR_NAMES = ['black', 'white'];
Grid.EMPTY_CHAR = '+';
Grid.DAME_CHAR = '?';
Grid.STONE_CHARS = '@O';
Grid.DEAD_CHARS = '&#';
Grid.TERRITORY_CHARS = '-:';
Grid.COLOR_CHARS = Grid.STONE_CHARS + Grid.DEAD_CHARS + Grid.TERRITORY_CHARS + Grid.DAME_CHAR + Grid.EMPTY_CHAR;
Grid.CIRCULAR_COLOR_CHARS = Grid.DAME_CHAR + Grid.EMPTY_CHAR + Grid.COLOR_CHARS;
Grid.ZONE_CODE = 100; // used for zones (100, 101, etc.); must be > COLOR_CHARS.length

// Possible values of a color (beside BLACK & WHITE)
Grid.EMPTY_COLOR = -1; // this is same as EMPTY, conveniently
Grid.DAME_COLOR = -2; // index of ? in above string; 2 from the end of the string
Grid.DEAD_COLOR = 2; // 2 and 3
Grid.TERRITORY_COLOR = 4; // 4 and 5

// Converts a "territory" character into an owner score (-1= black, +1= white)
// dame,empty, liveB,liveW deadB,deadW, terrB,terrW
Grid.territory2owner = [0,0, -1,+1, +1,-1, -1,+1];
// Converts potential territory number to a char (-1, -0.5, 0, +0.5, +1) -> char
Grid.territory2char = '-\'?.:';


Grid.prototype.init = function (initValue) {
    if (initValue === undefined) throw new Error('Grid init value must be defined');
    for (var j = this.gsize; j >= 1; j--) {
        var yxj = this.yx[j];
        for (var i = this.gsize; i >= 1; i--) {
            yxj[i] = initValue;
        }
    }
};

Grid.prototype.copy = function (sourceGrid) {
    if (sourceGrid.gsize !== this.gsize) throw new Error('Cannot copy between different sized grids');

    for (var j = this.gsize; j >= 1; j--) {
        var yxj = this.yx[j], srcYxj = sourceGrid.yx[j];
        for (var i = this.gsize; i >= 1; i--) {
            yxj[i] = srcYxj[i];
        }
    }
    return this;
};

/** Converts from goban grid (stones) to simple grid (colors)
 *  @param {Goban} goban - not modified
 *  @return {Grid} the grid (this)
 */
Grid.prototype.initFromGoban = function (goban) {
    var sourceGrid = goban.grid;
    if (sourceGrid.gsize !== this.gsize) throw new Error('Cannot copy between different sized grids');

    for (var j = this.gsize; j >= 1; j--) {
        var yxj = this.yx[j], srcYxj = sourceGrid.yx[j];
        for (var i = this.gsize; i >= 1; i--) {
            yxj[i] = srcYxj[i].color;
        }
    }
    return this;
};

// Returns the "character" used to represent a stone in text style
function colorToChar(color) {
    if (color === GRID_BORDER) return '(BORDER)';
    if (color >= Grid.ZONE_CODE) {
        return String.fromCharCode(('A'.charCodeAt() + color - Grid.ZONE_CODE));
    }
    if (color < Grid.DAME_COLOR || color >= Grid.COLOR_CHARS.length) {
        throw new Error('Invalid color ' + color);
    }
    if (color < 0) color += Grid.COLOR_CHARS.length;
    return Grid.COLOR_CHARS[color];
}

// Returns the name of the color/player (e.g. "black")
Grid.colorName = function (color) { // TODO remove me or?
    return Grid.COLOR_NAMES[color];
};

Grid.charToColor = function (char) {
    return Grid.CIRCULAR_COLOR_CHARS.indexOf(char) + Grid.DAME_COLOR;
};

function cell2char(c) {
    return colorToChar(typeof c === 'number' ? c : c.color);
}

Grid.prototype.toText = function (block) {
    return this.toTextExt(true, '\n', block || cell2char);
};

Grid.prototype.toLine = function (block) {
    return this.toTextExt(false, ',', block || cell2char);
};

// Receives a block of code and calls it for each vertex.
// The block should return a string representation.
// This method returns the concatenated string showing the grid.
Grid.prototype.toTextExt = function (withLabels, endOfRow, block) {
    var outYx = new Grid(this.gsize, '').yx;
    var maxlen = 1, i, j, val;
    for (j = this.gsize; j >= 1; j--) {
        for (i = 1; i <= this.gsize; i++) {
            val = block(this.yx[j][i]);
            if (val === null) continue;
            outYx[j][i] = val;
            maxlen = Math.max(maxlen, val.length);
        }
    }
    var numChar = maxlen;
    var white = '          ';
    var s = '';
    for (j = this.gsize; j >= 1; j--) {
        if (withLabels) s += '%2d'.format(j) + ' ';
        for (i = 1; i <= this.gsize; i++) {
            val = outYx[j][i];
            if (val.length < numChar) val = white.substr(1, numChar - val.length) + val;
            s += val;
        }
        s += endOfRow;
    }
    if (withLabels) {
        s += '   ';
        for (i = 1; i <= this.gsize; i++) {
            s += white.substr(1, numChar - 1) + Grid.xLabel(i);
        }
        s += '\n';
    }
    if (endOfRow !== '\n') s = s.chop(); // remove last endOfRow unless it is \n
    return s;
};

Grid.prototype.toString = function () {
    var s = '';
    for (var j = this.gsize; j >= 1; j--) {
        for (var i = 1; i <= this.gsize; i++) {
            s += colorToChar(this.yx[j][i]);
        }
        s += '\n';
    }
    return s;
};

// Returns a text "image" of the grid. See also copy? method.
// Image is upside-down to help compare with a copy paste from console log.
// So last row (j==gsize) comes first in image
Grid.prototype.image = function () {
    if (typeof this.yx[1][1] === 'object') {
        return this.toLine(function (s) {
            return colorToChar(s.color);
        });
    } else {
        return this.toLine(function (c) {
            return colorToChar(c);
        });
    }
};

// Watch out our images are upside-down on purpose (to help copy paste from screen)
// So last row (j==gsize) comes first in image
Grid.prototype.loadImage = function (image) {
    var rows = image.split(/\"|,/);
    if (rows.length !== this.gsize) {
        throw new Error('Invalid image: ' + rows.length + ' rows instead of ' + this.gsize);
    }
    for (var j = this.gsize; j >= 1; j--) {
        var row = rows[this.gsize - j];
        if (row.length !== this.gsize) throw new Error('Invalid image: row ' + row);
        for (var i = 1; i <= this.gsize; i++) {
            this.yx[j][i] = Grid.charToColor(row[i - 1]);
        }
    }
};

function reverseStr(s) {
    var res = '';
    for (var i = s.length - 1; i >= 0; i--) { res += s[i]; }
    return res;
}

Grid.flipImage = function (image) {
    return image.split(',').reverse().join();
};

Grid.mirrorImage = function (image) {
    return image.split(',').map(reverseStr).join();
};

Grid.flipAndMirrorImage = function (image) {
    return reverseStr(image);
};


var COLUMNS = 'abcdefghjklmnopqrstuvwxyz'; // NB: "i" is skipped

// Parses a move like "c12" into 3,12
// Throws OR returns null if move is not a vertex or illegal
Grid.move2xy = function (move, dontThrow) {
    var i = COLUMNS.indexOf(move[0]) + 1;
    var j = parseInt(move.substr(1, 2));
    if (!i || isNaN(j)) {
        if (dontThrow) return null;
        throw new Error('Illegal move parsed: ' + move);
    }
    return [i, j];
};

// Builds a string representation of a move (3,12->"c12")  
Grid.xy2move = function (i, j) {
    return COLUMNS[i - 1] + j;
};

// Converts a numeric X coordinate in a letter (e.g 3->c)
Grid.xLabel = function (i) {
    return COLUMNS[i - 1];
};

File: grid.rb
require_relative "stone_constants"

# A generic grid - a Goban owns a grid
class Grid

  COLOR_NAMES = ["black","white"]
  NOTATION_A = "a".ord # notation origin; could be A or a
  EMPTY_CHAR = "+"
  DAME_CHAR = "?"
  STONE_CHARS = "@O"
  DEAD_CHARS = "&#"
  TERRITORY_CHARS = "-:"
  COLOR_CHARS = STONE_CHARS+DEAD_CHARS+TERRITORY_CHARS+DAME_CHAR+EMPTY_CHAR
  EMPTY_COLOR = -1 # this is same as EMPTY, conveniently
  DAME_COLOR = -2 # index of ? in above string; 2 from the end of the string
  DEAD_COLOR = 2
  TERRITORY_COLOR = 4
  CIRCULAR_COLOR_CHARS = DAME_CHAR+EMPTY_CHAR+COLOR_CHARS
  ZONE_CODE = 100 # used for zones (100, 101, etc.); must be > COLOR_CHARS.size

  attr_reader :gsize, :yx

  def initialize(gsize=19)
    @gsize = gsize
    # TODO: use only 1 extra "nil" cell (0..gsize instead of 0..gsize+1)
    # Idea is to avoid to have to check i,j against gsize in many places.
    # In case of bug, e.g. for @yx[5][-1], Ruby returns you @yx[5][@yx.size] (looping back)
    # so having a real item (BORDER) on the way helps to detect a bug.
    @yx = Array.new(gsize+2) {Array.new(gsize+2,BORDER)}
  end

  def copy(source_grid)
    raise "Cannot copy between different sized grids" if source_grid.gsize != @gsize
    src_yx = source_grid.yx
    1.upto(@gsize) do |j|
      1.upto(@gsize) { |i| @yx[j][i] = src_yx[j][i] }
    end
    return self
  end

  # Converts from goban grid (stones) to simple grid (colors) REVIEWME
  def convert(source_grid)
    raise "Cannot copy between different sized grids" if source_grid.gsize != @gsize
    src_yx = source_grid.yx
    1.upto(@gsize) do |j|
      1.upto(@gsize) { |i| @yx[j][i] = src_yx[j][i].color }
    end
    return self
  end

  # Returns the "character" used to represent a stone in text style
  def self.color_to_char(color)
    return ("A".ord + color-ZONE_CODE).chr if color >= ZONE_CODE
    char = COLOR_CHARS[color]
    raise "Invalid color #{color}" if color<DAME_COLOR or color>=COLOR_CHARS.length
    return char
  end
  
  # Returns the name of the color/player (e.g. "black")
  def self.color_name(color) # TODO remove me or?
    return COLOR_NAMES[color]
  end

  def self.char_to_color(char)
    return CIRCULAR_COLOR_CHARS.index(char) + DAME_COLOR
  end

  # Receives a block of code and calls it for each vertex.
  # The block should return a string representation.
  # This method returns the concatenated string showing the grid.
  def to_text(with_labels=true, end_of_row="\n")
    yx = Grid.new(@gsize).yx
    maxlen = 1
    @gsize.downto(1) do |j|
      1.upto(@gsize) do |i|
        val = yield(@yx[j][i])
        val = "" if val == nil
        yx[j][i] = val
        maxlen = val.length if val.length > maxlen
      end
    end
    num_char = maxlen
    white = "          "
    s = ""
    @gsize.downto(1) do |j|
      s += "#{'%2d' % j} " if with_labels
      1.upto(@gsize) do |i|
        val = yx[j][i]
        val = white.slice(1, num_char-val.length) + val if val.length < num_char
        s += val
      end
      s += end_of_row
    end
    if with_labels
      s += "   "
      1.upto(@gsize) { |i| s += white.slice(1,num_char-1) + Grid.x_label(i) }
      s += "\n"
    end
    return s
  end

  def to_s
    s = ""
    @gsize.downto(1) do |j|
      1.upto(@gsize) { |i| s << Grid.color_to_char(@yx[j][i]) }
      s << "\n"
    end
    return s
  end

  # Returns a text "image" of the grid. See also copy? method.
  # Image is upside-down to help compare with a copy paste from console log.
  # So last row (j==gsize) comes first in image
  def image?
    if @yx[1][1].instance_of?(Stone) #FIXME
      return to_text(false, ",") { |s| Grid::color_to_char(s.color) } .chop
    else
      return to_text(false, ",") { |c| Grid::color_to_char(c) } .chop
    end
  end
  
  # Watch out our images are upside-down on purpose (to help copy paste from screen)
  # So last row (j==gsize) comes first in image
  def load_image(image)
    rows = image.split(/\"|,/)
    raise "Invalid image: #{rows.size} rows instead of #{@gsize}" if rows.size != @gsize
    @gsize.downto(1) do |j|
      row = rows[@gsize-j]
      raise "Invalid image: row #{row}" if row.length != @gsize
      1.upto(@gsize) do |i|
        @yx[j][i] = Grid.char_to_color(row[i-1])
      end
    end
  end

  # Parses a move like "c12" into 3,12
  def self.parse_move(move)
    return move[0].ord-NOTATION_A+1, move[1,2].to_i
  end
  
  # Builds a string representation of a move (3,12->"c12")  
  def self.move_as_string(col, row)
    return "#{(col+NOTATION_A-1).chr}#{row}"
  end
  
  # Converts a numeric X coordinate in a letter (e.g 3->c)
  def self.x_label(i)
    return (i+NOTATION_A-1).chr
  end

end
